research-software-directory/RSD-as-a-service
# 555 package managers

# Software maintainers can add software download/installation location(s)

Closes #555

Changes proposed in this pull request:
* Software maintainer can define "Package managers" to point to locations from where software can be installed.
* RSD will try to extract package manager "type" based on url provided and store it.
* RSD scrapers will try to extract additional stats (downloads and dependencies) from package manager api's when possible. Supported package managers are: anaconda, cran, dockerhub, github, maven, npm, pypi and other (not supported).
* Deep linking to software edit pages is implemented. The user can link to edit/contributors, edit organisations page etc,
* Drag and drop item resizing (scewness) is fixed and the ui responsiveness on position change is improved

How to test:
* `make start` to rebuild app and create test data
* login as rsd_user and create new software item,
* navigate to edit software, package managers page. Add number of locations: 
   * anaconda: https://anaconda.org/anaconda/spyder
   * CRAN: https://cran.r-project.org/web/packages/abglasso/index.html
   * dockerhub: https://hub.docker.com/r/swaggerapi/swagger-ui
   * maven: https://mvnrepository.com/artifact/junit/junit
   * npm: https://www.npmjs.com/package/npm
   * pypi: https://pypi.org/project/panda-model/
   * any other url will be defined as "other"
* navigate to edit contributors page. Add few contributors and confirm that position change of contributor item is working properly. Change contributor order and view software page.
* confirm deep linking is working by using back button from the software page. It should take you back to edit contributors page.
* validate package manager detection object (see below) includes correct host names (domains) for proper detection of package manager.

## Package manager detection

The detection is based on the host name (domain) provided in the URL. The `hostname` property of the object below should contain the host name provided in the URL.

```
{
  anaconda: {
    name: 'Anaconda',
    icon: '/images/anaconda-logo-96.svg',
    hostname: ['anaconda.org']
  },
  cran: {
    name: 'CRAN',
    icon: '/images/cran-r-logo.svg',
    hostname: ['cran.r-project.org']
  },
  dockerhub: {
    name: 'Dockerhub',
    icon: '/images/dockerhub-logo.webp',
    hostname: ['hub.docker.com']
  },
  maven: {
    name: 'Maven',
    icon: '/images/apache-maven-logo.svg',
    hostname: ['mvnrepository.com']
  },
  npm: {
    name: 'NPM',
    icon: '/images/npm-logo-64.png',
    hostname: ['www.npmjs.com','npmjs.com']
  },
  pypi: {
    name: 'PyPi',
    icon: '/images/pypi-logo.svg',
    hostname: ['pypi.org']
  }
}
```

## Example added package managers

![image](https://user-images.githubusercontent.com/9204081/227187915-be5cfca1-8670-4ee3-b202-5d4ec39a34ec.png)

PR Checklist:
